src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +672,477 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+            // Its important not to introduce suspension points in the loop
+            // below as it could invalidate the available_hwm_waiters iterator.
+            for (auto i : *available_hwm_waiters) {
+                const auto& req = ctx.requests[i];
+                requests.push_back(req);
+                results_map.push_back(i);
+
+                auto part = ctx.mgr.get(req.ktp());
+                if (!part) {
+                    results[i] = read_result(
+                      error_code::not_leader_for_partition);
+                    results[i].partition
+                      = ctx.requests[i].ktp().get_partition();
+                    co_return worker_result{
+                      .read_results = std::move(results),
+                      .errored_partitions = {i},
+                      .total_size = total_size,
+                    };
+                }
+                auto consensus = part->raft();
+                if (!consensus) {
+                    results[i] = read_result(
+                      error_code::not_leader_for_partition);
+                    results[i].partition
+                      = ctx.requests[i].ktp().get_partition();
+                    co_return worker_result{
+                      .read_results = std::move(results),
+                      .errored_partitions = {i},
+                      .total_size = total_size,
+                    };
+                }
+                last_visible_indexes[i] = consensus->last_visible_index();
+            }
+            available_hwm_waiters->clear();
+
+            auto new_results = co_await fetch_ntps_in_parallel(
+              ctx.mgr,
+              ctx.srv.get_replica_selector(),
+              std::move(requests),
+              ctx.foreign_read,
+              ctx.deadline,
+              ctx.bytes_left,
+              ctx.srv.memory(),
+              ctx.srv.memory_fetch_sem());
+
+            // Override the older results of the partitions just queried with
+            // the new results.
+            for (size_t i = 0; i < results_map.size(); i++) {
+                auto res_i = results_map[i];
+                auto& r = results[res_i];
+                total_size -= r.data_size_bytes();
+                r = std::move(new_results[i]);
+                total_size += r.data_size_bytes();
+                if (r.error != error_code::none) {
+                    errored_partitions.push_back(res_i);
+                }
+            }
+
+            if (
+              total_size > ctx.min_bytes || !errored_partitions.empty()
+              || ctx.as.abort_requested()) {
+                co_return worker_result{
+                  .read_results = std::move(results),
+                  .errored_partitions = std::move(errored_partitions),
+                  .total_size = total_size,
+                };
+            }
+
+            // Register new call backs for partitions just read.
+            for (auto res_i : results_map) {
```

Comment:
this loop can be factored out as it's similar to the one above?

Reply:
Good point, I should be able to factor out some of it.